internal/oidc/token/token_handler.go

```diff
@@ -165,3 +169,67 @@ func findOIDCProviderByNameAndValidateUID(
 	return nil, errorsx.WithStack(errUpstreamRefreshError.
 		WithHintf("Provider %q of type %q from upstream session data was not found.", s.ProviderName, s.ProviderType))
 }
+
+func upstreamLDAPRefresh(ctx context.Context, s *psession.CustomSessionData, providerCache oidc.UpstreamIdentityProvidersLister, username string, subject string) error {
+	// if you have neither a valid ldap session config nor a valid active directory session config
+	if (s.LDAP == nil || s.LDAP.UserDN == "") && (s.ActiveDirectory == nil || s.ActiveDirectory.UserDN == "") {
+		return errorsx.WithStack(errMissingUpstreamSessionInternalError)
+	}
+
+	// get ldap/ad provider out of cache
+	p, dn, err := findLDAPProviderByNameAndValidateUID(s, providerCache)
+	if err != nil {
+		return err
+	}
+	// run PerformRefresh
+	err = p.PerformRefresh(ctx, dn, username, subject)
+	if err != nil {
+		return errorsx.WithStack(errUpstreamRefreshError.WithHintf(
+			"Upstream refresh failed using provider %q of type %q.",
+			s.ProviderName, s.ProviderType).WithWrap(err))
+	}
+
+	return nil
+}
+
+func findLDAPProviderByNameAndValidateUID(
+	s *psession.CustomSessionData,
+	providerCache oidc.UpstreamIdentityProvidersLister,
+) (provider.UpstreamLDAPIdentityProviderI, string, error) {
+	var providers []provider.UpstreamLDAPIdentityProviderI
+	var dn string
+	if s.ProviderType == psession.ProviderTypeLDAP {
+		providers = providerCache.GetLDAPIdentityProviders()
+		dn = s.LDAP.UserDN
+	} else if s.ProviderType == psession.ProviderTypeActiveDirectory {
+		providers = providerCache.GetActiveDirectoryIdentityProviders()
+		dn = s.ActiveDirectory.UserDN
+	}
+
+	for _, p := range providers {
+		if p.GetName() == s.ProviderName {
+			if p.GetResourceUID() != s.ProviderUID {
+				return nil, "", errorsx.WithStack(errUpstreamRefreshError.WithHintf(
+					"Provider %q of type %q from upstream session data has changed its resource UID since authentication.",
+					s.ProviderName, s.ProviderType))
+			}
+			return p, dn, nil
+		}
+	}
+
+	return nil, "", errorsx.WithStack(errUpstreamRefreshError.
+		WithHintf("Provider %q of type %q from upstream session data was not found.", s.ProviderName, s.ProviderType))
```

Comment:
Maybe it's worth putting it in the cause instead of the hint, so users never see it but it makes it into the logs. It seems like it would be useful for the operator to know what happened.

Reply:
Resolved by putting in debug in d17403c